crates/crates_io_trustpub/src/keystore/impl.rs

```diff
@@ -0,0 +1,82 @@
+use super::OidcKeyStore;
+use super::load_jwks::load_jwks;
+use async_trait::async_trait;
+use jsonwebtoken::DecodingKey;
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+use tokio::sync::RwLock;
+
+/// The main implementation of the [`OidcKeyStore`] trait.
+///
+/// This struct fetches OIDC keys from a remote provider and caches them. If
+/// a key is not found in the cache, it will attempt to refresh the cached
+/// key set, unless the cache has just recently been refreshed.
+pub struct RealOidcKeyStore {
+    issuer_uri: String,
+    client: reqwest::Client,
+    cache: RwLock<Cache>,
+}
```

Comment:
This implementation roughly matches how PyPI caches the JWKS, though they save it in Redis with an expiration time, while we cache them in memory on each API server.

I hope the code comments make it clear enough how the implementation works.